src/client/java/dev/creesch/WebchatClient.java

```diff
@@ -39,16 +41,25 @@ private void handleMessage(Text message, MinecraftClient client) {
         }
 
         // Can't use GSON for Text serialization easily, using Minecraft's own serializer.
-        String minecraftChatJson = Text.Serialization.toJsonString(message, client.world.getRegistryManager());
+        JsonObject minecraftChatJsonObject = gson.fromJson(
+            Text.Serialization.toJsonString(message, client.world.getRegistryManager()),
+            JsonObject.class
+        );
         // Explicitly use UTC time for consistency across different timezones
         long timestamp = Instant.now(Clock.systemUTC()).toEpochMilli();
         ChatServerInfo serverInfo = MinecraftServerIdentifier.getCurrentServerInfo();
         String minecraftVersion = SharedConstants.getGameVersion().getName();
 
+        String playerName = client.player != null ? client.player.getName().getString() : "";
+        JsonObject messageObject = new JsonObject();
+        messageObject.addProperty("ping", isPing(message, playerName));
+        messageObject.add("component", minecraftChatJsonObject);
+        String jsonMessage = gson.toJson(messageObject);
```

Comment:
I feel like we might as well not serialize here and adjust `WebsocketJsonMessage.createChatMessage` to not take a string but an object.

Reply:
Okay, since I am working in the same area anyway (again, I'll handle merge conflicts). This is what I landed on for now since I do need the String JSON for storage and UUID generation anyway I moved a few things 

```java 
    private void handleMessage(Text message, MinecraftClient client) {
        if (client.world == null) {
            return;
        }

        // Can't use GSON for Text serialization easily, using Minecraft's own serializer.
        String minecraftChatJson = Text.Serialization.toJsonString(message, client.world.getRegistryManager());

        // Explicitly use UTC time for consistency across different timezones
        long timestamp = Instant.now(Clock.systemUTC()).toEpochMilli();
        ChatServerInfo serverInfo = MinecraftServerIdentifier.getCurrentServerInfo();
        String minecraftVersion = SharedConstants.getGameVersion().getName();
        // UUID used to prevent duplicates when doing 
        String messageUUID = UUID.nameUUIDFromBytes((timestamp + minecraftChatJson).getBytes()).toString();

        // Back to objects we go
        JsonObject messageObject = new JsonObject();
        messageObject.addProperty("history", false);
        messageObject.addProperty("uuid", messageUUID);
        messageObject.add("component", gson.fromJson(
            minecraftChatJson,
            JsonObject.class
        ));

        WebsocketJsonMessage chatMessage = WebsocketJsonMessage.createChatMessage(
                timestamp,
                serverInfo,
                messageObject,
                minecraftVersion
        );

        // And back to JSON string again
        String jsonChatMessage = gson.toJson(chatMessage);

        LOGGER.info(jsonChatMessage);
        // TODO: do storage somewhere around here
        webInterface.broadcastMessage(jsonChatMessage);
    }
```